File: src/lexer/source_line.cpp
#include "lexer/token.h"
#include "lexer/source_line.h"

SourceLine::SourceLine(nat num)
: locale({0, 0}), num(num), string(), marks() {}

void SourceLine::ignore() {
    locale[0] = locale[1];
}

void SourceLine::mark(Token& token) {
    marks.push_back(token.locale);
}

void SourceLine::increment() { locale[1] += 1; }

Locale SourceLine::newLocale() {
    Locale result(locale);
    locale[0] = locale[1];

    return result;
}

str SourceLine::unreadString() const {
    return string.substr(locale[1]);
}

str SourceLine::raw() const {
    nat col1 = 0, col2 = 0;

    for(const Locale& locale : marks) {
        if(col1 == 0 or locale[0] < col1) col1 = locale[0];
        if(col2 == 0 or locale[1] > col2) col2 = locale[1];
    }

    str line;

    for(nat i = 0; i < string.size(); i++) {
        line += (i < col1 or i > col2) ? ' ' : '^';
    }

    return fmt::format("{:4d}|{}    |{}", num, string, line);
}

File: app/src/main/kotlin/sea/lexer/SourceLine.kt
package sea.lexer

import sea.lexer.Locale
import sea.lexer.Token

data class SourceLine(val num: Int, val string: String) {
    private var locale = Locale()
    val marks = ArrayList<Locale>()

    val unreadString: String
        get() = string.substring(locale[1])

    fun ignore() { locale[0] = locale[1] }
    fun mark(locale: Locale) { marks.add(locale) }
    fun mark(token: Token) { marks.add(token.locale) }
    fun increment() { if(locale[1] < string.length) locale[1] += 1 }

    fun newLocale(): Locale {
        val result = locale.copy()
        locale[0] = locale[1]
        return result
    }

    fun raw(): String {
        var col1 = -1
        var col2 = -1

        for(locale in marks) {
            println(locale)
            if(col1 == -1 || locale[0] < col1) col1 = locale[0]
            if(col2 == -1 || locale[1] > col2) col2 = locale[1]
        }

        var underline = ""

        for(i in string.indices) {
            underline += if(i < col1 || i >= col2) ' ' else '^'
        }

        val line = if(string == "") "EOF\n" else string
        if(string == "") underline = "^^^"

        return "${"%4d".format(num)}|$line    |$underline"
    }
}
